airflow/providers/amazon/aws/sensors/s3.py

```diff
@@ -78,27 +80,32 @@ def __init__(
     ):
         super().__init__(**kwargs)
         self.bucket_name = bucket_name
-        self.bucket_key = bucket_key
+        self.bucket_key = [bucket_key] if isinstance(bucket_key, str) else bucket_key
         self.wildcard_match = wildcard_match
         self.aws_conn_id = aws_conn_id
         self.verify = verify
         self.hook: Optional[S3Hook] = None
 
-    def _resolve_bucket_and_key(self):
+    def _resolve_bucket_and_key(self, key):
         """If key is URI, parse bucket"""
         if self.bucket_name is None:
-            self.bucket_name, self.bucket_key = S3Hook.parse_s3_url(self.bucket_key)
+            return S3Hook.parse_s3_url(key)
         else:
-            parsed_url = urlparse(self.bucket_key)
+            parsed_url = urlparse(key)
             if parsed_url.scheme != '' or parsed_url.netloc != '':
                 raise AirflowException('If bucket_name provided, bucket_key must be relative path, not URI.')
+            return self.bucket_name, key
 
-    def poke(self, context: 'Context'):
-        self._resolve_bucket_and_key()
-        self.log.info('Poking for key : s3://%s/%s', self.bucket_name, self.bucket_key)
+    def _key_exists(self, key):
+        bucket_name, key = self._resolve_bucket_and_key(key)
+        self.log.info('Poking for key : s3://%s/%s', bucket_name, key)
         if self.wildcard_match:
-            return self.get_hook().check_for_wildcard_key(self.bucket_key, self.bucket_name)
-        return self.get_hook().check_for_key(self.bucket_key, self.bucket_name)
+            return self.get_hook().check_for_wildcard_key(key, bucket_name)
+
+        return self.get_hook().check_for_key(key, bucket_name)
+
+    def poke(self, context: 'Context'):
+        return all(self._key_exists(key) for key in self.bucket_key)
```

Comment:
> What will happen if I check for 300 keys? If I get it right this will make 300 API calls per poke? If so I think it's not good.

I agree but I also need to be backward compatible. That's the current behavior in `S3PrefixSensor` so I cannot really change this logic. I cannot either use some list APIs because the scope of these APIs are per bucket, and I can imagine providing 300 different buckets

> Also if I poke for bucket_key=['a','b'] and 'a' is present but 'b' is not why should I keep poke for a?

That's a good point. I can look into a way to optimize that

Reply:
> Also if I poke for bucket_key=['a','b'] and 'a' is present but 'b' is not why should I keep poke for a?

Bouncing back on this question. Actually, having file 'a' present at time N, it does not mean file 'a' is still present at time N+1. So I dont think we can really optimize here. Example: we are waiting for file 'a' and 'b' in a same bucket.
- File 'a' gets created
- File 'a' gets removed
- File 'b' gets created

In this situation, the sensor should never stop waiting. If we go down your optimization path, the sensor will act as though the 2 files are present

Reply:
@vincbeck valid point.
As for the API calls... we preserve backward compatibility in terms of not breaking user code.
If there is a way to optimize (convert 300 API calls to 1 call without user changing anything) then we can do it.
I'm not sure if it's doable - this is what I wanted to find out.

Reply:
The only way to optimize it would be to use `ListObjects` instead but there are some caveats.
1. `ListObjects` works for a given bucket. So If you want check 300 different keys, all of them in different bucket, it would make the problem worse
2. It would make the code more complex because by using `ListObjects` you need to handle the different pages, making the necessary checks that the file is returned as part of `ListObjects`
3. By looking really quick on how current users use `S3PrefixSensor` (https://github.com/search?q=S3PrefixSensor&type=code), most of it are single file usage

Because of these points, in my opinion, it's not worth using `ListObjects` instead of `HeadObject`

Reply:
Thanks! make sense.
Should we mention 1-2 sentences about this in the operator docs?
Just to notify that the sensor was not designed for high volume of keys

Reply:
Totally! That's a good point. Let me add that

Reply:
Done